### ISSUE TITLE
Less results than real Google

### ISSUE THREAD

**Issue by WAZAAAAA0** [searxng/searxng#2438]
*(continuation of https://github.com/searx/searx/issues/1236)*
*(cross-posted in https://github.com/benbusby/whoogle-search/issues/1004, https://github.com/hnhx/librex/issues/225)*

It's been years. Google-searching through a "privacy search engine frontend" will rarely find as many results as the real Google.

Here's a simple test to verify that: come up with a unique Google query that will find as few results as possible, preferably not in English. For example in my test I used `"sfendazi"` but you might need your own unique query since the results come and go. Perform the same search on every public instance, and observe how many find the same results (if the results contain garbage unrelated stuff, consider it a failure). This was the outcome yesterday as of 2023-05-15:

<details>
  <summary><a href="https://github.com/hnhx/librex#instances">LibreX instances:</a> 20 tested, 0 work</summary>

lmao
</details>
<details>
  <summary><a href="https://github.com/benbusby/whoogle-search#public-instances">Whoogle instances:</a> 17 tested, 3 work</summary>

https://s.tokhmi.xyz
https://whoogle.dcs0.hu
https://whoogle.privacydev.net
</details>
<details>
  <summary><a href="https://searx.space/">SearX/SearXNG instances:</a> 92 tested, 19 work if you tweak a setting, only 1 works with defaults</summary>

(the only one that works with defaults is https://opnxng.com)
https://priv.au
https://xo.wtf
https://offtheradar.info
https://searx.oakleycord.dev
https://searx.cthd.icu
https://ooglester.com
https://search.bus-hit.me
https://myprivatesrx.us
https://coppedge.info
https://search.neet.works
https://search.zzls.xyz
https://search.us.projectsegfau.lt
https://s.frlt.one
https://searx.sev.monster
https://stalk.antelope.day
https://searx.esmailelbob.xyz
https://search.serginho.dev
https://search.cronobox.one
https://searx.mxchange.org
</details>

Those 19 instances I listed think they're "smart" and have set their `Search language` to `[auto]`, which auto-selects it based on your browser headers... or they're simply set to something arbitrary, like `[en-US]`. Choosing `[all]` fixes the problem for them.
Meanwhile, the rest of the instances somehow will not find the correct results even when set to `[all]`. From what I've tested with a local SearXNG instance, adding search query parameter `nfpr=1` (along with the pre-existing `safe=off` and `filter=0`) to `searxng/searx/engines/google.py` fixed it. Here's what they do:
- `nfpr=1` -> Showing results for XXX Search instead for YYY ON
- `safe=off` -> SafeSearch OFF
- `filter=0` -> Include omitted results ON

Changing the `Interface language` is fine. Actually, I'd argue language auto-detection should happen to the interface, not to the search results filter, which would be consistent with how major search engines work.

Honestly, just take the `Search language` option away, it does more harm than good. Or at least make `[all]` the default and lock the option behind huge warning signs with flaming skulls that searching will be seriously degraded for everyone if anything other than `[all]` is selected. People don't understand this is the equivalent setting they're touching (taken from Google's official [advanced search page](https://www.google.ca/advanced_search?hl=en&authuser=0)):
![](https://i.imgur.com/qmrKqwO.png)

# TL;DR
Here's a picture to sum up the problem most search frontends are facing:
![](https://i.imgur.com/F6HWkDR.png)
Proposed fixes:
1. remove `Search language` and default it to `[all]`
2. give `[auto]` to the `Interface language` instead
3. add these 3 parameters to unlock all the possible Google results `?safe=off&nfpr=1&filter=0`

**Comment by xieve** [comment]
Here's the thing: if I set default language to all, some answers will be in English, even if I search in another language. (E.g. Wikipedia instant answers) The better solution would be to set the language for those search engines that deliver significantly worse results with the language filter to all, and leave the option in for those that don't.

Edit: Sadly, it seems that specifying `language: all` for a search engine is not valid. Is there another way to disable locales for one engine specifically without editing source code?

**Comment by return42** [comment]
> Edit: Sadly, it seems that specifying `language: all` for a search engine is not valid. Is there another way to disable locales for one engine specifically without editing source code?

No, this is not possible.  This would not fit into the concept of selecting a language and/or locale for a search engine.

The search engines support different languages and locales:
 
--> https://github.com/searxng/searxng/blob/master/searx/data/engine_traits.json

The  user selects one search language (de) or even one region (de-AT) aka "locale" and then SearXNG has to determine the best matching locale for each search engine.

https://github.com/searxng/searxng/blob/80aaef6c95b572df1fa3a8c30b7fdc1538d7a306/searx/locales.py#L228-L376

The concept has been implemented in:

- https://github.com/searxng/searxng/pull/2269

In this commit message https://github.com/searxng/searxng/commit/9ae409a05a its described in more details and the comments in the test cases explain what it means in practice:

https://github.com/searxng/searxng/blob/80aaef6c95b572df1fa3a8c30b7fdc1538d7a306/tests/unit/test_locales.py#L60-L111

**Comment by JamesC1** [comment]
I am affected by this fault/feature.

I object to the characterisation `The user selects one search language` above.

After I have made my first search, there is a language drop-down on the results page and the characterisation becomes true enough.  Working backwards from there:
- before I make the first search, there is no visible language setting; it is hidden in Settings/<drop down>/OK
- there is no obvious URL query/parameter/fragment to set the default language
- if I find instances from https://searx.space, there is no drop-down there to set the language in the engine URL

My current work-around is to do one search for a sarcastic word related to the above, then edit the search text, then choose a language (almost always "all"), which will trigger the first real search.

Are there good reasons related to the design of SearxNG, why my three bullet-points could not be implementable features?

**Comment by return42** [comment]
@JamesC1 your question is OT / if you want to discuss this topic please open a new discussion.

---
But to answer your questions ..

What initial language is used depends on the setup of an engine / by default SearXNG uses the language the browser sends .. this is also what common search provider like google or ddg do.

If this is not what the user want's (the browser language), the user can set the language in the preferences.

Alternatively the user can use the [search syntax](https://docs.searxng.org/user/search-syntax.html#select-language)  

Language alias 'all' is most often a bad choice since not every search engine has such an alias and the *meaning* off *all* can be different from engine to engine.

**Comment by return42** [comment]
> Less results than real Google

AFAIK this has been fixed in ..

- https://github.com/searxng/searxng/pull/2673

**Comment by unixfox** [comment]
@WAZAAAAA0 could you take a look again at this new improvement by re-running your tests? thank you

**Comment by WAZAAAAA0** [comment]
@return42 @unixfox situation didn't improve.

- Google
**5 real results** that DO contain the exact searched word (4 in Greek, 1 in English, Italian IP, no cookies)
SAMPLE URL: https://www.google.ca/search?q=%22sfendaki%22&safe=off&nfpr=1&filter=0
SNAPSHOT: https://i.imgur.com/7RKq9yH.png

- Several updated SearX instances (darmarit.org, searxng.nicfab.eu, search.sapti.me, opnxng.com, priv.au, searx.daetalytica.io)
**1 full page of wrong results** that DO NOT contain the searched word
SAMPLE URL: https://searx.daetalytica.io/search?language=all&time_range=&safesearch=0&categories=general&q=!go+%22sfendaki%22
SNAPSHOT: https://i.imgur.com/ebebD92.png

To be honest I have long since switched to **eTools.ch** for my meta-search engine needs. It needs some minor adjustments to be serviceable (uBO filters to reduce the clutter, a Redirector rule to avoid affiliate links) and it doesn't support Images/Videos/News etc. categories, or date-filters, but at least it's the most solid one. I had no idea they've been around since 2000 wow.
